### PR TITLE
arch/risc-v: Add support for S-mode flat build 

### PR DESCRIPTION
**Pull request by inochisa**
## Summary
There is no need to use kernel build for S-mode all the time. As cpu
scratch is supported for non kernel mode build, it is possible to use
flat build for S-mode.

Add flat support for risc-v S mode build and new example for this build `flats` and `flats64`

## Impact
`k230` and `qemu-rv` SoC

## Testing
`rv-virt:flats` and `rv-virt:flats64`

log with the following test command (for 32-bit, replace 64 with 32 in the command line):
```
qemu-system-riscv64 \
  -monitor stdio -d guest_errors -D nuttx-qemu.log -semihosting \
  -M virt,aclint=on -cpu rv64 -m 2G \
  -chardev vc,id=char0,logfile=nuttx.log,signal=off \
  -serial chardev:char0 \
  -kernel nuttx
```
[nuttx-flats.log](https://github.com/user-attachments/files/15814346/nuttx-flats.log)
[nuttx-flats64.log](https://github.com/user-attachments/files/15814348/nuttx-flats64.log)

Regression test log for  `knsh32` and `knsh64`:
[nuttx-knsh32.log](https://github.com/user-attachments/files/15814218/nuttx-knsh32.log)
[nuttx-knsh64.log](https://github.com/user-attachments/files/15814222/nuttx-knsh64.log)

### CODE REVIEW

**Review by acassis — APPROVED** [review]

**Comment by lupyuen** [comment]
@yf13 Could you review the K230 changes? And maybe test on K230? Thanks!

**Comment by lupyuen** [comment]
@inochisa I think this code might potentially impact `knsh32` and `knsh64`? Could you do the Regression Testing and attach the NuttX Logs for these targets? Thanks!

**Comment by yf13** [comment]
@inochisa, can you teach what is the difference between this build mode and the KERNEL mode? The `rv-virt/smp64_s` looks like a FLAT build that can run in S-mode of RISCV? 

My stock 10.2 toolchain can't build `rv-virt/smp64_s` directly but can build it with the following changes on your version:
```
-CONFIG_ARCH_CHIP_QEMU_RV_ISA_V=y
+CONFIG_DEBUG_FEATURES=y
-CONFIG_LIBM_TOOLCHAIN=y
+CONFIG_LIBM=y
```

For K230, do you have any configs want me to try?

**Comment by yf13** [comment]
@inochisa, can you teach what is the difference between this build mode and the KERNEL mode? The `rv-virt/smp64_s` looks like a FLAT build that can run in S-mode of RISCV? 

My stock 10.2 toolchain can't build `rv-virt/smp64_s` directly but can build it with the following changes on your version:
```
-CONFIG_ARCH_CHIP_QEMU_RV_ISA_V=y
+CONFIG_DEBUG_FEATURES=y
-CONFIG_LIBM_TOOLCHAIN=y
+CONFIG_LIBM=y
```
The build can boot into NSH prompt with my stock 6.2 qemu which uses OpenSBI-1.0 firmware with `-smp 8 -kernel nuttx` command options.

Have you tried this on K230 device? or do you have any particular points or configs want me to try on K230? as you may not that K230 doesn't seem to have SMP.

**Comment by inochisa** [comment]
> @inochisa, can you teach what is the difference between this build mode and the KERNEL mode? The `rv-virt/smp64_s` looks like a FLAT build that can run in S-mode of RISCV?

Yes, it is just a flat build that can run in S-mode

> 
> My stock 10.2 toolchain can't build `rv-virt/smp64_s` directly but can build it with the following changes on your version:
> 

This is pretty weird, I just generated this config file from the `smp64` config, The only change is just `ARCH_USE_S_MODE` and change uart irq number.

> ```
> -CONFIG_ARCH_CHIP_QEMU_RV_ISA_V=y
> +CONFIG_DEBUG_FEATURES=y
> -CONFIG_LIBM_TOOLCHAIN=y
> +CONFIG_LIBM=y
> ```
> 
> The build can boot into NSH prompt with my stock 6.2 qemu which uses OpenSBI-1.0 firmware with `-smp 8 -kernel nuttx` command line.
> 
> Have you tried this on K230 device? or do you have any particular points or configs want me to try on K230? as you may note that K230 doesn't seem to have SMP.
> 

For test, you can just use `nsh` config and enable `ARCH_USE_S_MODE` for it. And it is also needed to change the irq number for S-mode. I think this change have no side affect but my K230 board is now out of my hand....

> As previously K230 port followed the original QEMU port, so you updated K230 as its so simialr to that of QEMU?

You are right. I just check the related code and changed it.

**Comment by inochisa** [comment]
> Here I renamed my tweaked config as `rv-virt/flat4s` which means `FLAT for S-mode`. Maybe you want drop the SMP and vector stuffs so that it is as simple as `nsh64` and name the new config as `rv-virt/snsh64` which means `S-mode` variant of the `nsh64` so that it is easier for people to try.

You are right, it is better to keep this simple. I will took a look and get rid of any unnecessary options.

**Comment by inochisa** [comment]
Update:
1. replace `non-kernel` -> `flat`
2. add simple `flats` and `flats64` for basic 32/64 bit configuration. and drop the complex `smp64_s`.
3. add regression test log

**Comment by lupyuen** [comment]
Hmmm the RISC-V Build failed with a download error, lemme restart the build...

https://github.com/apache/nuttx/actions/runs/9494382489/job/26164701349?pr=12488

```text
====================================================================================
Configuration/Tool: rv-virt/netnsh_smp
------------------------------------------------------------------------------------
  Cleaning...
  Configuring...
  Building NuttX...
28) Failed to connect to github.com port 443 after 134197 ms: Connection timed out
make[1]: *** [libmetal.defs:77: libmetal.zip] Error 28
```

**Comment by lupyuen** [comment]
@inochisa The build is OK now. Could you update the above 2 fixes for the docs? So I can approve the PR. Thanks!

**Comment by inochisa** [comment]
> @inochisa The build is OK now. Could you update the above 2 fixes for the docs? So I can approve the PR. Thanks!

Do you mean the doc for the two new configuration? I have already updated it, but forgot to mention. If there are others needed to update, please let me know.

**Comment by lupyuen** [comment]
@inochisa Hmmm sorry I don't see the updates, wonder if I missed something. Here are the changes we need:

(1) From https://github.com/inochisa/nuttx/blob/nommu-s-mode/Documentation/platforms/risc-v/qemu-rv/boards/rv-virt/index.rst

Please change both `S mode`:

```text
If testing with S mode build, remove the ``-bios none`` option. S mode build
```

To `S-mode` (both of them):

```text
If testing with S-mode build, remove the ``-bios none`` option. S-mode build
```

(2) Also from https://github.com/inochisa/nuttx/blob/nommu-s-mode/Documentation/platforms/risc-v/qemu-rv/boards/rv-virt/index.rst

Please change both `S mode`:

```text
flats: Similar to the `nsh`_ configuration, but running in S mode.
flats64: Similar to the `nsh`_ configuration, but running in S mode.
```

To `S-mode` (both of them):

```text
flats: Similar to the `nsh`_ configuration, but running in S-mode.
flats64: Similar to the `nsh`_ configuration, but running in S-mode.
```

(3) From https://github.com/inochisa/nuttx/blob/nommu-s-mode/arch/risc-v/Kconfig

Please change `Run the NuttX`:

```text
bool "Run the NuttX in S-mode"
```

To `Run NuttX`

```text
bool "Run NuttX in S-mode"
```

Thanks!

File: Documentation/platforms/risc-v/qemu-rv/boards/rv-virt/index.rst [hunk 1]
```diff
@@ -334,6 +337,18 @@ smp64
 Similar to the `nsh`_ configuration, but with SMP support
 This configuration is used for 64-bit RISC-V
 
+flats
+-------
+
+Similar to the `nsh`_ configuration, but running in S mode.
```
Comment (lupyuen) [inline Documentation/platforms/risc-v/qemu-rv/boards/rv-virt/index.rst:343]:
```suggestion
Similar to the `nsh`_ configuration, but running in S-mode.
```

File: Documentation/platforms/risc-v/qemu-rv/boards/rv-virt/index.rst [hunk 2]
```diff
@@ -334,6 +337,18 @@ smp64
...
+This configuration is used for 32-bit RISC-V
+
+flats64
+-------
+
+Similar to the `nsh`_ configuration, but running in S mode.
```
Comment (lupyuen) [inline Documentation/platforms/risc-v/qemu-rv/boards/rv-virt/index.rst:349]:
```suggestion
Similar to the `nsh`_ configuration, but running in S-mode.
```

**Review by lupyuen — CHANGES_REQUESTED** [review]
Here are the proposed fixes thanks!

**Comment by inochisa** [comment]
> Here are the proposed fixes thanks!

Updated, thanks.

File: Documentation/platforms/risc-v/qemu-rv/boards/rv-virt/index.rst [hunk 3]
```diff
@@ -63,13 +63,16 @@ Finally, to run it, use the following command:
...
 And, for 64-bit configurations::
 
-    $ qemu-system-riscv64 -semihosting -M virt,aclint=on -cpu rv64 -smp 8 -bios none -kernel nuttx -nographic
+    $ qemu-system-riscv64 -semihosting -M virt,aclint=on -cpu rv64 -smp <cpu number> -bios none -kernel nuttx -nographic
 
-If testing with kernel build, remove the ``-bios none`` option. Kernel build
+``-smp`` option can be only used in smp build, and the ``cpu number`` needs
+to be set to the same value as ``CONFIG_SMP_NCPUS`` in the build config file.
+
+If testing with S mode build, remove the ``-bios none`` option. S-mode build
```
Comment (lupyuen) [inline Documentation/platforms/risc-v/qemu-rv/boards/rv-virt/index.rst:75]:
```suggestion
If testing with S-mode build, remove the ``-bios none`` option. S-mode build
```

**Review by lupyuen — CHANGES_REQUESTED** [review]
Sorry there's one more `S-mode` here thanks!

File: Documentation/platforms/risc-v/qemu-rv/boards/rv-virt/index.rst (hunk 3 above)
Comment (inochisa) [reply Documentation/platforms/risc-v/qemu-rv/boards/rv-virt/index.rst]: Fixed, thanks.

**Review by lupyuen — APPROVED** [review]
Looks great, thank you so much :-)

**Review by jerpelea — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**